Fix event capturing in beforeCreate lifecycle hook

wtho opened the pull request:
The event capturing was installed in the wrapper constructor, after the component (vm) was created and mounted, which is too late to catch events emitted in lifecycle hooks like `beforeCreate`.

Two additional tests were written, in `emitted.spec.js` and `emittedByOrder.spec.js`, and an existing one was adjusted. Notice I had to treat Vue v2.0.8 differently to pass the compatibility tests (in `emittedByOrder.spec.js`), as that version emits lifecycle hooks events, whereas the others do not.

Implementation:
* A simple event captor object is created when `mount` is called before the `vm` is created.
* The captor gets passed to the `createInstance` method, where it will get installed as a mixin on the used Vue instance. The installation uses the same method as the wrapper used before. The vm creation continues as usual
* When the wrapper gets created in the `mount` method, the captor gets passed as well. Instead of installing the capturing on its own, the wrapper will use the event captor object.
* Flow type definitions were added for the EventCaptor, as it is passed in two methods.

This solves #215.

eddyerburgh commented on src/wrappers/vue-wrapper.js:
```diff
@@ -24,9 +23,9 @@ export default class VueWrapper extends Wrapper implements BaseWrapper {
+    this._emittedByOrder = options.eventCaptor.emittedByOrder
 
-    logEvents(vm, this._emitted, this._emittedByOrder)
+    // logEvents(vm, this._emitted, this._emittedByOrder)
```
We can delete this line

eddyerburgh commented on src/lib/create-instance.js:
```diff
@@ -13,14 +13,23 @@ import extractOptions from '../options/extract-options'
 
+  vue.mixin({
+    beforeCreate: function () {
+      logEvents(this, eventCaptor.emitted, eventCaptor.emittedByOrder)
```
I actually think we should use `this.__emitted` and `this.__emittedBy`, because otherwise (I think) all events logged by any of the child components will be added to the eventCaptor. Using `this.__emitted` means we can support child component emitted, which won't work at the moment.

wtho replied:
Yeah, I think you are right, if a `localVue` instance will be reused, the captor might be added several times. I thought it might be better to keep it non-intrusive, but I will do it using `this.__emitted` then.

wtho replied:
> Using this.__emitted means we can support child component emitted, which won't work at the moment.

How would that work? Is there anything required to implement this as well? Right now this does not get captured by `wrapper.emitted()`:
```js
    const wrapper = mount({
      template: '<child-component></child-component>',
      components: {
        'child-component': {
          beforeCreate: function () {
            this.$emit('childEvent')
          }
        }
      }
    })
```

eddyerburgh replied:
Sorry, that wasn't very clear. I meant that at the moment (in this PR) only root components will support emitted, child components won't, because when the VueWrapper is created with them, they won't have an eventCaptor to pass in. 

This should work:
```js
    const wrapper = mount({
      template: '<child-component></child-component>',
      components: {
        'child-component': {
          beforeCreate: function () {
            this.$emit('childEvent')
          }
        }
      }
    })
wrapper.find(childComponent).vm.emitted('childEvent') // returns array

Thinking about the underscore, I think it's fine to have a single underscore.

eddyerburgh commented on src/wrappers/vue-wrapper.js:
```diff
@@ -24,9 +23,9 @@ export default class VueWrapper extends Wrapper implements BaseWrapper {
-    this._emitted = Object.create(null)
-    this._emittedByOrder = []
+    this._emitted = options.eventCaptor.emitted
+    this._emittedByOrder = options.eventCaptor.emittedByOrder
```
We should add this._emitted (maybe this.__emitted to be safe) in the mixin, then every component will have them added. Could you write a test to make sure this is true? Every child component should have a clean emitted object that only refers to events emitted by itself

wtho commented:
Applied the changes. `this.__emitted` and `this.__emittedByOrder` is used to capture the events. These two fields will then be used by the wrapper, when it is created and they will not be removed.

I removed previously added type definitions in flow, that were not required anymore.

I created an `addEventLogger` method in `logEvents.js` to keep `createInstance.js` clean. It violates the `add-something.js`-pattern, as it does not use the default export, but I prefer it being with the `logEvents` method (that is nowhere else used at the moment than in `addEventLogger`).

I added one more test to check for side effects on localVue with several wrappers / components. The events emitted by children are not captured by the parent, but on the child component. I might add another test to show this.

eddyerburgh commented:
Thanks so much @wtho  🙂
